src/roiextractors/extractors/memmapextractors/memmapextractors.py

```diff
@@ -64,43 +62,80 @@ def get_num_channels(self):
         """
         imaging = imaging_extractor
-        video_to_save = np.memmap(
+        file_size_in_bytes = Path(imaging.file_path).stat().st_size
+        available_memory_in_bytes = psutil.virtual_memory().available
+        buffer_size_in_bytes = buffer_gb * 10**9
+        if available_memory_in_bytes < buffer_size_in_bytes:
+            raise f"Not enough memory available memory {available_memory_in_bytes* 10**9} for buffer size {buffer_gb}"
+
+        num_frames = imaging.get_num_frames()
+        memmap_shape = imaging.video_structure.build_video_shape(n_frames=num_frames)
+        dtype = imaging.get_dtype()
+
+        # Load the memmap
+        video_memmap = np.memmap(
             save_path,
-            shape=(
-                imaging.get_num_frames(),
-                imaging.get_num_channels(),
-                imaging.get_image_size()[0],
-                imaging.get_image_size()[1],
-            ),
-            dtype=imaging.get_dtype(),
+            shape=memmap_shape,
+            dtype=dtype,
             mode="w+",
         )
 
-        if verbose:
-            for ch in range(imaging.get_num_channels()):
-                print(f"Saving channel {ch}")
-                for i in tqdm(range(imaging.get_num_frames())):
-                    plane = imaging.get_frames(i, channel=ch)
-                    video_to_save[ch, i] = plane
-        else:
-            for ch in range(imaging.get_num_channels()):
-                for i in range(imaging.get_num_frames()):
-                    plane = imaging.get_frames(i, channel=ch)
-                    video_to_save[ch, i] = plane
+        if file_size_in_bytes < buffer_size_in_bytes:
+            video_data_to_save = imaging.get_frames()
+            video_memmap[:] = video_data_to_save
 
-        video_to_save.flush()
+        else:
+            chunk_size_in_bytes = int(buffer_size_in_bytes)
+            type_size = np.dtype(dtype).itemsize
+
+            n_channels = imaging.get_num_channels()
+            pixels_per_frame = n_channels * np.product(imaging.get_image_size())
+            bytes_per_frame = type_size * pixels_per_frame
+            frames_per_chunk = chunk_size_in_bytes // bytes_per_frame
+
+            # If the chunk size is smaller than one image, the iterator goes over one image only.
+            frames_per_chunk = max(frames_per_chunk, 1)
+            iterator = range(0, num_frames, frames_per_chunk)
+            if verbose:
+                iterator = tqdm(iterator, ascii=True, desc="Writing to .dat file")
+
+            for frame in iterator:
+                end_frame = min(frame + frames_per_chunk, num_frames)
+
+                # Get the video chunk
+                video_chunk = imaging.get_video(start_frame=frame, end_frame=end_frame)
+
+                # Fit the video chunk in the memmap array
+                indices = np.arange(start=frame, stop=end_frame)
+                axis_to_expand = (
+                    imaging.video_structure.rows_axis,
```

Comment:
Other `ImagingExtractors` don't have `video_structure` as an attribute besides the new `NumpyMemmapExtractor`, right? These static methods ought to be designed to apply to any other extractor, especially if they're caching a binary blob.

Reply:
This is done in the https://github.com/catalystneuro/roiextractors/pull/125.